atef/widgets/ophyd.py

```diff
@@ -0,0 +1,729 @@
+            ...
+
+    def clear(self):
+        """Clear all models and reset the device."""
+        self.stop()
+        for model in self.models.values():
+            model.stop()
+
+        self.models.clear()
+        self._device = None
+
+    def _table_context_menu(self, pos: QtCore.QPoint) -> None:
+        """Context menu for the table."""
+        self.menu = QtWidgets.QMenu(self)
+        index: QtCore.QModelIndex = self.indexAt(pos)
+        if index is not None:
+            def copy(*_):
+                copy_to_clipboard(index.data())
+
+            def select_attr(*_):
+                self.attributes_selected.emit([row_data.attr])
+
+            if index.data() is not None:
+                copy_action = self.menu.addAction(f"&Copy: {index.data()}")
+                copy_action.triggered.connect(copy)
+
+            row_data = self.get_data_from_proxy_index(index)
+            if row_data is not None:
+                select_action = self.menu.addAction(
+                    f"&Select attribute: {row_data.attr}"
+                )
+                select_action.triggered.connect(select_attr)
+
+        self.menu.exec_(self.mapToGlobal(pos))
+
+    def get_data_from_proxy_index(
+        self, index: QtCore.QModelIndex
+    ) -> Optional[OphydAttributeData]:
+        """Proxy model index -> source model index -> OphydAttributeData."""
+        model = self.current_model
+        if model is None:
+            return None
+
+        return model.get_data_for_row(self.proxy_model.mapToSource(index).row())
+
+    @property
+    def current_model(self) -> Optional[PolledDeviceModel]:
+        """The current device model."""
+        try:
+            return self.models[self._device]
+        except KeyError:
+            return None
+
+    @property
+    def device(self) -> Optional[ophyd.Device]:
+        """The currently-configured ophyd Device."""
+        return self._device
+
+    @device.setter
+    def device(self, device: Optional[ophyd.Device]):
+        if device is self._device:
+            return
+
+        self.stop()
+
+        self._device = device
+        if not device:
+            return
+
+        try:
+            model = self.models[device]
+        except KeyError:
+            model = PolledDeviceModel(device=device, poll_rate=self.poll_rate)
+            self.models[device] = model
+            new_model = True
+        else:
+            new_model = False
+
+        model.poll_rate = self.poll_rate
+        model.data_updates_started.connect(self.data_updates_started.emit)
+        model.data_updates_finished.connect(self.data_updates_finished.emit)
+        self.proxy_model.setSourceModel(model)
+
+        if new_model:
+            # Only start an update if we haven't previously gotten information
+            # about the device
+            model.start()
+
+
+class OphydDeviceTableWidget(DesignerDisplay, QtWidgets.QFrame):
+    """
+    A convenient frame with an embedded OphydDeviceTableView.
```

Comment:
What exactly is this used for over just having an OphydDeviceTableView?

Reply:
`OphydDeviceTableView` is _just_ the table view, but this widget `OphydDeviceTableWidget` includes the buttons and filters that surround it. 

The goal in my head was to allow piece-meal combining of the table view into new widgets, but also give a conveniently-instantiable widget that does probably everything you'd want.